### PR TITLE
Refactor session auth to use cookie jar and share base service http client

### PR DESCRIPTION
**Pull request by eiri**
## PR summary

This refactoring makes CouchDB Session authenticator to use cookiejar to store authsession cookie and set it to requests and makes `BaseService` to set its HTTP client as client for CouchDB Session authenticator to share its cookiejar.

This is motivated by changes in CouchDB where server now sets auth session cookie on basic auth.

## PR Checklist

Please make sure that your PR fulfills the following requirements:

- [x] The commit message follows the
[Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type  

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: auth/couchdb_session_authenticator.go
Comment (ricellis) [inline auth/couchdb_session_authenticator.go:1]: copyright update needed

File: auth/couchdb_session_authenticator_test.go
Comment (ricellis) [inline auth/couchdb_session_authenticator_test.go:1]: copyright

File: auth/session.go
Comment (ricellis) [inline auth/session.go:1]: copyright

File: auth/session_test.go
Comment (ricellis) [inline auth/session_test.go:1]: copyright

File: common/cloudant_base.go
Comment (ricellis) [inline common/cloudant_base.go:1]: copyright

File: common/cloudant_base_test.go
Comment (ricellis) [inline common/cloudant_base_test.go:1]: copyright

**Review by ricellis — CHANGES_REQUESTED** [review]
In general looks good, but a few nits, linting things and tidy ups.

File: auth/couchdb_session_authenticator.go [hunk 1]
```diff
@@ -116,23 +119,45 @@ func (a *CouchDbSessionAuthenticator) Validate() error {
...
+	return err
+}
+
+// SetClient sets the http client for the authenticator. It injects a cookiejar
+// if supplied client is missing one.
+func (a *CouchDbSessionAuthenticator) SetClient(client *http.Client) error {
+	var err error
+	if client.Jar == nil {
+		client.Jar, err = cookiejar.New(&cookiejar.Options{PublicSuffixList: publicsuffix.List})
 	}
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go]: Well, in base this is happening to base's http client and here this is done to auth's http client.

It's a bit messy, really - we have to keep client setter public, so base could use it, because this is a separate module and we can't make http client into a required field on auth initiator, because of mapenv constructor.

So if we are allowing setting an http client and want things to work we either inject cookiejar transparently, same as we do in base service, or returning an error if it is not set and then remove creation of default http client in authenticator initiator. I prefer to inject to have the logic in-line with base service and honestly, because this is a minimal change for a price of a bit of code duplication.

Another approach could be to move the whole thing into `common` and make client field private, but that's bigger refactoring and a breaking change.

File: auth/couchdb_session_authenticator.go [hunk 2]
```diff
@@ -116,23 +119,45 @@ func (a *CouchDbSessionAuthenticator) Validate() error {
...
+	if err := a.refreshCookie(); err != nil {
+		return nil, err
+	}
+	for _, cookie := range a.Client.Jar.Cookies(url) {
+		if cookie.Name == "AuthSession" {
+			return cookie, nil
+		}
+	}
+	return nil, fmt.Errorf("Missing AuthSession cookie")
 }
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go]: done in https://github.com/IBM/cloudant-go-sdk/pull/399/commits/28f2bc75fadea028f09097d910853cbc58497742

File: auth/couchdb_session_authenticator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ricellis) [reply auth/couchdb_session_authenticator.go]: This concern is resolved by the newer changes.

File: auth/couchdb_session_authenticator.go [hunk 3]
```diff
@@ -116,23 +115,30 @@ func (a *CouchDbSessionAuthenticator) Validate() error {
...
 
-	request.AddCookie(cookie)
-	return nil
+	return err
 }
 
-// getCookie returns an AuthSession cookie to be used in a request.
+// SetClient sets the http client for the authenticator. It injects a cookiejar
+// if supplied client is missing one.
+func (a *CouchDbSessionAuthenticator) SetClient(client *http.Client) {
```
Comment (ricellis) [inline auth/couchdb_session_authenticator.go:135]: I think this function should set the "inherited" value `DisableSSLVerification`.

Actually thinking about it more that is a property of the "service" and not the `http.Client` so this isn't the correct place to read the new value and set it. Another complication is that the `URL` is also inhereted. Currently we declare:
>// CouchDB URL inherited from the service config.
and
>// A flag that indicates whether verification of the server's SSL certificate should be disabled; INherired from the service config

Looking at the code I think that was possibly only ever true if the authenticator was init'd from `GetAuthenticatorFromEnvironment` and never reflected any subsequent changes caused by calls to either `SetHttpClient` or `SetServiceURL`. I think this remains true at the current state of this PR.

I don't see there is any value now in exposing `DisableSSLVerfication` as it will always be doing whatever the `SetClient` client has been configured to do (and that in turn will pretty much always be the base service client).
`URL` is a little more complicated because we will need access to set that. So we can either leave it exposed or internalize it as well and add a `SetURL` similar to `SetClient`. In either case the value will need to be updated from the `SetServiceURL` override.

**Review by ricellis — CHANGES_REQUESTED** [review]
This looks good, but a couple of niggles remain.

File: auth/couchdb_session_authenticator.go (hunk 3 above)
Comment (eiri) [reply auth/couchdb_session_authenticator.go:135]: I have a commit ["fix: put back apparently required auth field"](https://github.com/IBM/cloudant-go-sdk/pull/399/commits/0cc08ba21762bb454fe5daa80f1eded1f1c44b23) that says it all 😄 

This is basically part of core's support for `PROPNAME_SVC_DISABLE_SSL` and we use it as a public var in our base service override, so can't be changed or removed.

> I think this function should set the "inherited" value DisableSSLVerification.

I'm not so sure. I used to set it on constructed http client, because it could come from env, but now we are getting this for free from base service and are treating `SetClient` as "do whatever you want" function. It'd be weird to override one arbitrary parameter in programmatical setup based on value from env.

It probably should be used in `NewCouchDbSessionAuthenticator` directly instead, but given that we are almost immediately throwing that client away I feel this is more hustle than it worth it. 


`URL` is more complicated, because it is a property of request, not http client and we reconstruct it from request's URL evey time.

Basically both `URL` and `DisableSSLVerification`  are part of Authenticator interface, though implicit ones, rather than real meaningful vars of session authenticator.

Comment (ricellis) [reply auth/couchdb_session_authenticator.go:135]: Alright, I can live with the disable SSL stcking around and the flag potentially being inconsistent with the current client (my hatred of that particular option is well known and anyone using it has bigger problems than this IMO)!

I'm ok with `URL` as public, but I do still think that `SetServiceURL` should update the session authenticator URL. If I follow the pattern to get a new default instance (which points to `http://localhost:5984`) and then call:
`service.SetServiceURL('https://myservice.example')` I'd want it to be trying to auth against `https://myservice.example/_session` not `https://localhost:5984/_session`.

Comment (eiri) [reply auth/couchdb_session_authenticator.go:134]: Well, that's not a concern, because we are [composing URL](https://github.com/IBM/cloudant-go-sdk/blob/main/auth/couchdb_session_authenticator.go#L120) from request on each authenticate call, so `auth.URL` is more of a container to consequent calls to `_session` after that, it's never going to be out of sync. Honestly we have it as public for the same reason as `DisableSSLVerification` - those are "hidden" authenticator interface expectations.

I did this anyway in https://github.com/IBM/cloudant-go-sdk/pull/399/commits/0749c50ba1d8f962b63173838b500c19777245be just for completion sake.

File: auth/couchdb_session_authenticator.go [hunk 4]
```diff
@@ -116,23 +115,30 @@ func (a *CouchDbSessionAuthenticator) Validate() error {
...
 	}
 
-	request.AddCookie(cookie)
-	return nil
+	return err
 }
 
-// getCookie returns an AuthSession cookie to be used in a request.
+// SetClient sets the http client for the authenticator. It injects a cookiejar
+// if supplied client is missing one.
```
[the remark this answers is not in the document]
Comment (eiri) [reply auth/couchdb_session_authenticator.go]: I removed this and the rest of "inherited" notes in https://github.com/IBM/cloudant-go-sdk/pull/399/commits/06cfe86830d19855ab9774f084437b93fefba40d

File: auth/couchdb_session_authenticator.go (hunk 3 above)
Comment (ricellis) [reply auth/couchdb_session_authenticator.go:134]: Ah a subtle deviation from what the other implementations do that I had overlooked as one of the pieces that isn't actually changing here!
Thanks for closing the loop though.
I guess the per request `URL` code isn't strictly required any more, but I think it can stay (particularly to help in those weird potential 2 client scenarios we've discused).

**Review by ricellis — APPROVED** [review]

File: auth/couchdb_session_authenticator.go (hunk 3 above)
Comment (eiri) [reply auth/couchdb_session_authenticator.go:134]: Exactly. We can remove it now, just set auth's URL on client inject and on `SetServiceURL`, but then that "you can set your client" scenario is becoming a bit more awkward, because user will additionally have to set URL too or it'll stick to whatever was set there by default. On the other hand this is what user is probably want,  so I can't decide one way or another.

File: common/cloudant_base.go [hunk 5]
```diff
@@ -165,21 +162,48 @@ func (c *BaseService) Request(req *http.Request, result interface{}) (detailedRe
...
-			if serviceUrlPathSegmentsSize == 1 && serviceUrlPathSegments[0] == "" {
-				c.serviceUrlPathSegmentsSize = 0
-			} else {
-				c.serviceUrlPathSegmentsSize = serviceUrlPathSegmentsSize
-			}
-		}
+	if err != nil {
+		return err
+	}
+	// Set CouchDb Session's uth URL to Base service URL
```
Comment (emlaver) [inline common/cloudant_base.go:168]:
```suggestion
	// Set CouchDb Session's auth URL to Base service URL
```

**Review by emlaver — APPROVED** [review]
